flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/StandaloneCompletedCheckpointStore.java

```diff
@@ -77,8 +79,32 @@ public void addCheckpoint(CompletedCheckpoint checkpoint) throws Exception {
 	}
 
 	@Override
-	public CompletedCheckpoint getLatestCheckpoint() {
-		return checkpoints.isEmpty() ? null : checkpoints.getLast();
+	public CompletedCheckpoint getLatestCheckpoint(boolean isPreferCheckpointForRecovery) {
+		if (checkpoints.isEmpty()) {
+			return null;
+		}
+
+		CompletedCheckpoint candidate = checkpoints.getLast();
```

Comment:
Do you think we could somehow avoid having pretty much exactly the same logic duplicated twice (ZK/Standalone)? Maybe the whole logic could be implemented in the CompletedCheckpointStore interface based on getAllCheckpoints()

Reply:
Agree, actually, I found the similar code snippet when I was implementing. The difference between the original implementation of the `getLatestCheckpoint` is the usage of `peekLast` and `getLast`. I was afraid to destroy the original semantics. If it does not matter, I'd like to do this refactor.

Reply:
we should preserve the current behaviour, but the difference between getLast and peekLast is minor, just check for empty list I guess

Reply:
I will give a default implementation in the `CompletedCheckpointStore `:

```java
default CompletedCheckpoint getLatestCheckpoint(boolean isPreferCheckpointForRecovery) throws Exception {
		if (getAllCheckpoints().isEmpty()) {
			return null;
		}

		CompletedCheckpoint candidate = getAllCheckpoints().get(getAllCheckpoints().size() - 1);
		if (isPreferCheckpointForRecovery && getAllCheckpoints().size() > 1) {
			List<CompletedCheckpoint> allCheckpoints;
			try {
				allCheckpoints = getAllCheckpoints();
				ListIterator<CompletedCheckpoint> listIterator = allCheckpoints.listIterator(allCheckpoints.size() - 1);
				while (listIterator.hasPrevious()) {
					CompletedCheckpoint prev = listIterator.previous();
					if (!prev.getProperties().isSavepoint()) {
						candidate = prev;
						LOG.info("Found a completed checkpoint before the latest savepoint, will use it to recover!");
						break;
					}
				}
			} catch (Exception e) {
				LOG.error("Method getAllCheckpoints caused exception : ", e);
				throw new FlinkRuntimeException(e);
			}
		}

		return candidate;
	}
```

Reply:
Sounds good, as long as we can be sure getAllCheckpoints() returns checkpoint in order.

Reply:
Does it mean that the latest checkpoint is be surely  in the last position in list? thanks

Reply:
@R-CodeBoy Yes, the `getAllCheckpoints ` returns an ArrayList which wrapped an ArrayDeque.

Reply:
@yanghua Do you compare the cost time of recovery between checkpoint and savepoint,In my options ,sometimes it depends on the checkpoint interval when considering the recovery speed or reprocessing the data.

Reply:
Actually, it is a config option which depends on users themselves production scene. Some scene, the cost of rolling offset back is higher than recovery time, however, some scene is converse. To guess the scene is difficult to a general compute engine.